pkg/collector/listeners/kubelet.go

```diff
@@ -0,0 +1,217 @@
+	Ports         []int
+	Pid           int
+}
+
+func init() {
+	Register("kubelet", NewKubeletListener)
+}
+
+func NewKubeletListener() (ServiceListener, error) {
+	watcher, err := kubelet.NewPodWatcher(5 * time.Second)
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to kubelet, Kubernetes listener will not work: %s", err)
+	}
+	return &KubeletListener{
+		watcher:  watcher,
+		services: make(map[ID]Service),
+		ticker:   time.NewTicker(5 * time.Second),
+		stop:     make(chan bool),
+	}, nil
+}
+
+func (l *KubeletListener) Listen(newSvc chan<- Service, delSvc chan<- Service) {
+	// setup the I/O channels
+	l.newService = newSvc
+	l.delService = delSvc
+
+	go func() {
+		for {
+			select {
+			case <-l.stop:
+				return
+			case <-l.ticker.C:
+				// Compute new/updated pods
+				updatedPods, err := l.watcher.PullChanges()
+				if err != nil {
+					log.Error(err)
+					continue
+				}
+				for _, pod := range updatedPods {
+					// Ignore pending/failed/succeeded/unknown states
+					if pod.Status.Phase == "Running" {
+						l.processNewPod(pod)
+					}
+				}
+				// Compute deleted pods
+				expiredContainerList, err := l.watcher.ExpireContainers()
+				if err != nil {
+					log.Error(err)
+					continue
+				}
+				for _, containerID := range expiredContainerList {
+					l.removeService(ID(containerID))
+				}
+			}
+		}
+	}()
+}
+
+func (l *KubeletListener) Stop() {
+	l.ticker.Stop()
+	l.stop <- true
+}
+
+func (l *KubeletListener) processNewPod(pod *kubelet.Pod) {
+	for _, container := range pod.Status.Containers {
+		l.createService(ID(container.ID), pod)
+	}
+}
+
+func (l *KubeletListener) createService(id ID, pod *kubelet.Pod) {
+	svc := PodContainerService{
+		ID: id,
+	}
+	podName := pod.Metadata.Name
+
+	// AD Identifiers
+	var containerName string
+	for _, container := range pod.Status.Containers {
+		if container.ID == string(svc.ID) {
+			svc.ADIdentifiers = append(svc.ADIdentifiers, container.Name)
+			long, short, _, err := docker.SplitImageName(container.Image)
+			if err != nil {
+				log.Warnf("Error while spliting image name: %s", err)
+			}
+			if len(long) > 0 {
+				svc.ADIdentifiers = append(svc.ADIdentifiers, long)
+			}
+			if len(short) > 0 && short != long && short != container.Name {
+				svc.ADIdentifiers = append(svc.ADIdentifiers, short)
+			}
+			containerName = container.Name
+			break
+		}
+	}
+
+	// Hosts
+	podIp := pod.Status.PodIP
+	if podIp == "" {
+		log.Errorf("Unable to get pod %s IP", podName)
+	}
+	svc.Hosts = map[string]string{"pod": podIp}
+
+	// Ports
+	var ports []int
+	for _, container := range pod.Spec.Containers {
+		if container.Name == containerName {
+			for _, port := range container.Ports {
+				ports = append(ports, port.ContainerPort)
+			}
+			break
+		}
+	}
+	svc.Ports = ports
+	if len(svc.Ports) == 0 {
+		// Port might not be specified in pod spec
+		log.Errorf("Failed to get ports for pod %s", podName)
+	}
+
+	// Tags
+	_, err := svc.GetTags()
```

Comment:
Let's skip that and let `GetTags` be called if the service actually matches a template. It will be called by the templateresolver.